src/backend/libs-and-metadata.md

```diff
@@ -0,0 +1,192 @@
+# Libraries and Metadata
+
+When the compiler sees a reference to an external crate, it needs to load some
+information about that crate. This chapter gives an overview of that process,
+and the supported file formats for crate libraries.
+
+## Libraries
+
+A crate dependency can be loaded from an `rlib`, `dylib`, or `rmeta` file. A
+key point of these file formats is that they contain `rustc`-specific
+[*metadata*](#metadata). This metadata allows the compiler to discover enough
+information about the external crate to understand the items it contains,
+which macros it exports, and *much* more.
+
+### rlib
+
+An `rlib` is an [archive file], which is similar to a tar file. This file
+format is specific to `rustc`, and may change over time. This file contains:
+
+* Object code, which is the result of code generation. This is used during
+  regular linking. There is a separate `.o` file for each [codegen unit]. The
+  codegen step can be skipped with the [`-C
+  linker-plugin-lto`][linker-plugin-lto] CLI option, which means each `.o`
```

Comment:
Why does rustc reuse .o as the extension for bitcode files? I would expect .ll or something.

Reply:
What do you mean? My understanding is this talking about object code (not LLVM bitcode), so `.o` is correct.

Reply:
See the line below, this is reused for bitcode too: https://github.com/rust-lang/rustc-dev-guide/pull/1044/files#diff-3a3e84f49881b6db90d4538e80369c04098f7a6fe6d20cca0ccad9e46fead011R24

Reply:
Ah, it confused me that your comment only extended until the line before :)

Reply:
In the past, the bitcode was stored as a separate, compressed file in the archive (with the `.bc.z` extension).  This was changed to simplify the implementation and to match the behavior of clang, which always places the bitcode in the `.o` file.

Reply:
Kind of like `.o` for "output" :)